internal/resources/cloudprovider/utils_test.go

```diff
@@ -0,0 +1,162 @@
+package cloudprovider_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/grafana/terraform-provider-grafana/v3/internal/common"
+	"github.com/grafana/terraform-provider-grafana/v3/internal/common/cloudproviderapi"
+	"github.com/grafana/terraform-provider-grafana/v3/internal/testutils"
+	"github.com/stretchr/testify/require"
+)
+
+type testConfig struct {
+	stackID   string
+	accountID string
+	roleARN   string
+}
+
+func makeTestConfig(t require.TestingT) testConfig {
+	// Uses a pre-existing account resource so that we don't need to create a new one for every test run
+	accountID := os.Getenv("GRAFANA_CLOUD_PROVIDER_TEST_AWS_ACCOUNT_RESOURCE_ID")
+	require.NotEmpty(t, accountID, "GRAFANA_CLOUD_PROVIDER_TEST_AWS_ACCOUNT_RESOURCE_ID must be set")
+
+	roleARN := os.Getenv("GRAFANA_CLOUD_PROVIDER_AWS_ROLE_ARN")
+	require.NotEmpty(t, roleARN, "GRAFANA_CLOUD_PROVIDER_AWS_ROLE_ARN must be set")
+
+	stackID := os.Getenv("GRAFANA_CLOUD_PROVIDER_TEST_STACK_ID")
+	require.NotEmpty(t, stackID, "GRAFANA_CLOUD_PROVIDER_TEST_STACK_ID must be set")
+
+	// Make sure the account exists and matches the role ARN we expect for testing
+	client := testutils.Provider.Meta().(*common.Client).CloudProviderAPI
+	gotAccount, err := client.GetAWSAccount(context.Background(), stackID, accountID)
+	require.NoError(t, err)
+	require.Equal(t, roleARN, gotAccount.RoleARN)
+
+	return testConfig{
+		stackID:   stackID,
+		accountID: accountID,
+		roleARN:   roleARN,
+	}
+}
+
+func regionsString(regions []string) string {
+	b := new(bytes.Buffer)
+	for _, region := range regions {
+		fmt.Fprintf(b, "\n\t\t\"%s\",", region)
+	}
+	fmt.Fprintf(b, "\n\t")
+	return b.String()
+}
+
+func servicesString(svcs []cloudproviderapi.AWSCloudWatchService) string {
+	if len(svcs) == 0 {
+		return ""
+	}
+	b := new(bytes.Buffer)
+	for _, svc := range svcs {
+		fmt.Fprintf(b, "\n\t\t")
+		fmt.Fprintf(b, `{
+			name = "%[1]s",
+			metrics = [%[2]s],
```

Comment:
arent we using this notation for metrics, or are both supported:
```
    metric {
      name        = "CPUUtilization"
      statistics  = ["Average"]
    }

    metric {
      name        = "StatusCheckFailed"
      statistics  = ["Maximum"]
    }
```

Reply:
the purposes of these util functions are to convert the API structs into a string encoding [Terraform's object format](https://developer.hashicorp.com/terraform/language/expressions/type-constraints#object) that can be plugged into the `for_each` block in e.g. https://github.com/grafana/terraform-provider-grafana/blob/439626c56f114eca1f8b216d980e18b15c20130d/internal/resources/cloudprovider/resource_aws_cloudwatch_scrape_job_test.go#L308